pallets/omnipool/src/lib.rs

```diff
@@ -0,0 +1,1459 @@
+			&(&asset_out_state).into(),
+			amount,
+			T::AssetFee::get(),
+		)
+		.ok_or(ArithmeticError::Overflow)?;
+
+		ensure!(
+			*state_changes.asset.delta_reserve >= limit,
+			Error::<T>::BuyLimitNotReached
+		);
+
+		let new_asset_out_state = asset_out_state
+			.delta_update(&state_changes.asset)
+			.ok_or(ArithmeticError::Overflow)?;
+
+		// Token updates
+		T::Currency::transfer(
+			T::HubAssetId::get(),
+			who,
+			&Self::protocol_account(),
+			*state_changes.asset.delta_hub_reserve,
+		)?;
+		T::Currency::transfer(
+			asset_out,
+			&Self::protocol_account(),
+			who,
+			*state_changes.asset.delta_reserve,
+		)?;
+
+		// Imbalance update
+		let current_imbalance = <HubAssetImbalance<T>>::get();
+		Self::update_imbalance(current_imbalance, state_changes.delta_imbalance)?;
+
+		Self::set_asset_state(asset_out, new_asset_out_state);
+
+		Self::deposit_event(Event::SellExecuted {
+			who: who.clone(),
+			asset_in: T::HubAssetId::get(),
+			asset_out,
+			amount_in: *state_changes.asset.delta_hub_reserve,
+			amount_out: *state_changes.asset.delta_reserve,
+		});
+
+		Ok(())
+	}
+
+	/// Swap asset for Hub Asset
+	/// Special handling of buy trade where asset in is Hub Asset.
+	fn buy_asset_for_hub_asset(
+		who: &T::AccountId,
+		asset_out: T::AssetId,
+		amount: Balance,
+		limit: Balance,
+	) -> DispatchResult {
+		ensure!(
+			HubAssetTradability::<T>::get().contains(Tradability::SELL),
+			Error::<T>::NotAllowed
+		);
+
+		let asset_out_state = Self::load_asset_state(asset_out)?;
+
+		ensure!(
+			asset_out_state.tradable.contains(Tradability::BUY),
+			Error::<T>::NotAllowed
+		);
+
+		let state_changes = hydra_dx_math::omnipool::calculate_buy_for_hub_asset_state_changes(
+			&(&asset_out_state).into(),
+			amount,
+			T::AssetFee::get(),
+		)
+		.ok_or(ArithmeticError::Overflow)?;
+
+		ensure!(
+			*state_changes.asset.delta_reserve <= limit,
+			Error::<T>::SellLimitExceeded
+		);
+
+		let new_asset_out_state = asset_out_state
+			.delta_update(&state_changes.asset)
+			.ok_or(ArithmeticError::Overflow)?;
+
+		T::Currency::transfer(
+			T::HubAssetId::get(),
+			who,
+			&Self::protocol_account(),
+			*state_changes.asset.delta_hub_reserve,
+		)?;
+		T::Currency::transfer(
+			asset_out,
+			&Self::protocol_account(),
+			who,
+			*state_changes.asset.delta_reserve,
+		)?;
+
+		let current_imbalance = <HubAssetImbalance<T>>::get();
+		Self::update_imbalance(current_imbalance, state_changes.delta_imbalance)?;
+
+		Self::set_asset_state(asset_out, new_asset_out_state);
+
+		Self::deposit_event(Event::BuyExecuted {
+			who: who.clone(),
+			asset_in: T::HubAssetId::get(),
+			asset_out,
+			amount_in: *state_changes.asset.delta_hub_reserve,
+			amount_out: *state_changes.asset.delta_reserve,
+		});
+
+		Ok(())
+	}
+
+	/// Buy hub asset from the pool
+	/// Special handling of buy trade where asset out is Hub Asset.
+	fn buy_hub_asset(_who: &T::AccountId, _asset_in: T::AssetId, _amount: Balance, _limit: Balance) -> DispatchResult {
```

Comment:
this should be documented in buy/sell function

Reply:
it is mentioned in the buy / sell that buy/sell is special case. but it does not require to add the extrisnic docs part. it is the same math, there is nothing really different. 

the reason why you need to handle this separate because you deal with one asset state only and not two ( if you trade 2 other assets). 
that means, you load one asset state only. but math is the same only simplified, some parts are removed as no need to deal with second asset state.

Reply:
it should be documented that you cannot buy hub asset - every outcome (including errors) should be covered by documentation

Reply:
it is already there that asset trade state of asset in must include SELL flag and asset out must include BUY flag.

Reply:
so how can user tell beforehand that this wont be allowed?

Reply:
when FE loads asset state - there is the information what possible to do with asset (sell,buy,add/remove liquid), so it should display the information and allow/not allow an operation.